### DISCUSSION TITLE
How to create new docks on the fly while preserving across saves?

### DISCUSSION THREAD

**Discussion by i-make-robots** [andrewauclair/ModernDocking#240] [Q&A]
I apologize in advance, this question might take a few tries to clear up.  I'm able to create docks with hard-coded UIDs and have them preserve from one session to another.  Suppose I have a factory that creates new docks and panels.  How might I preserve the panels across saves?  

One example is that I have "details" tab A that appears when I select an item.  I'd like to be able to Pin it and then select a second item.  This would create details tab B with a randomly selected UID.  When I save and reload I have a random UID and I'm not sure how to determine what dock B is supposed to contain.

**Comment by tlf30** [comment]
I'll give an example of what I have done. I have tabs that open files, I make the uid in the following format `<tool>-<file>`. An example would be `msglog-C:\some\path\log.json` Then when the system loads the tab, I can use the uid to see what I needed to open. 

Another option I have done is save a parameter in the tab with the more complex metadata needed for what to restore back into the tab. So the tab will have a uid of `<tool>-<random>` but there are saved parameters in the tab that tell the application what to load into it. 

Both methods work well.

**Reply by i-make-robots** [reply]
saved parameters?  first I've heard of it.  What part of the API is that?

**Reply by tlf30** [reply]
https://github.com/andrewauclair/ModernDocking/blob/df195635351600ca1c55104ea2b21e3f76428ddc/docking-api/src/ModernDocking/api/AppStateAPI.java#L252

See the discussion here: https://github.com/andrewauclair/ModernDocking/discussions/111

Basically, you can persist custom properties with each dockable. This is done using setProperty and getProperty functions from the AppStateAPI.

**Reply by i-make-robots** [reply]
What listeners are you using to create the correct instance of `tool` on restore?

**Comment by andrewauclair** [comment]
Dynamic dockables are supported. You can create dockables on the fly and ModernDocking will persist and reload them by remembering what class they were. When reloading the layout, ModernDocking will create an instance of the class and restore any properties for it, like @tlf30 mentions. I've been meaning to write a specific example of how this works because the current demo does a poor job of it.
